crates/pathfinder/src/state/merkle_tree.rs

```diff
@@ -1651,6 +1733,544 @@ mod tests {
+        use stark_hash::StarkHash;
+
+        impl EdgeProofNode {
+            fn hash(&self) -> StarkHash {
+                // Code taken from [merkle_node::EdgeNode::calculate_hash]
+                let child_hash = self.child_hash;
+
+                // Path should be valid, so `unwrap()` is safe to use here.
+                let path = StarkHash::from_bits(&self.path).unwrap();
+                let mut length = [0; 32];
+                // Safe as len() is guaranteed to be <= 251
+                length[31] = self.path.len() as u8;
+
+                // Length should be smaller than the maximum size of a stark hash.
+                let length = StarkHash::from_be_bytes(length).unwrap();
+
+                stark_hash::stark_hash(child_hash, path) + length
+            }
+        }
+
+        impl BinaryProofNode {
+            fn hash(&self) -> StarkHash {
+                // Code taken from [merkle_node::EdgeNode::calculate_hash]
+                stark_hash::stark_hash(self.left_hash, self.right_hash)
+            }
+        }
+
+        impl ProofNode {
+            fn hash(&self) -> StarkHash {
+                match self {
+                    ProofNode::Binary(bin) => bin.hash(),
+                    ProofNode::Edge(edge) => edge.hash(),
+                }
+            }
+        }
+
+        #[derive(Debug, PartialEq)]
+        pub enum Membership {
+            Member,
+            NonMember,
+        }
+
+        /// Verifies that the key `key` with value `value` is indeed part of the MPT that has root
+        /// `root`, given `proofs`.
+        /// Supports proofs of non-membership as well as proof of membership: this function returns
+        /// an enum corresponding to the membership of `value`, or returns `None` in case of a hash mismatch.
+        /// The algorithm follows this logic:
+        /// 1. init expected_hash <- root hash
+        /// 2. loop over nodes: current <- nodes[i]
+        /// 1. verify the current node's hash matches expected_hash (if not then we have a bad proof)
+        /// 2. move towards the target - if current is:
+        ///    1. binary node then choose the child that moves towards the target, else if
+        ///    2. edge node then check the path against the target bits
+        ///       1. If it matches then proceed with the child, else
+        ///       2. if it does not match then we now have a proof that the target does not exist
+        /// 3. nibble off target bits according to which child you got in (2). If all bits are gone then you have reached the target and the child hash is the value you wanted and the proof is complete.
+        /// 4. set expected_hash <- to the child hash
+        fn verify_proof(
+            root: StarkHash,
+            key: &BitSlice<Msb0, u8>,
+            value: StarkHash,
+            proofs: &Vec<ProofNode>,
+        ) -> Option<Membership> {
+            // Protect from ill-formed keys
+            if key.len() != 251 {
```

Comment:
btw @Mirko-von-Leipzig is there a constant defined somewhere that corresponds to this ? (251)

Reply:
The tree itself has a `max_height` field. It should be a const but we anticipated needing different heights for other trees.